bug fix: valid but different LockPoints after a reorg

glozow opened the pull request:
I introduced a bug in #22677 (sorry! :sweat_smile:)

Mempool entries cache `LockPoints`, containing the first height/blockhash/`CBlockIndex*` at which the transaction becomes valid. During a reorg, we re-check timelocks on all mempool entries using `CheckSequenceLocks(useExistingLockPoints=false)` and remove any now-invalid entries. `CheckSequenceLocks()` also mutates the `LockPoints` passed in, and we update valid entries' `LockPoints` using `update_lock_points`. Thus, `update_lock_points(lp)` needs to be called right after `CheckSequenceLocks(lp)`, otherwise we lose the data in `lp`. I incorrectly assumed they could be called in separate loops.

The incorrect behavior introduced is: if we have a reorg in which a timelocked mempool transaction is still valid but becomes valid at a different block, the cached `LockPoints` will be incorrect.

This PR fixes the bug, adds a test, and adds an assertion at the end of `removeForReorg()` to check that all mempool entries' lockpoints are valid. You can reproduce the bug by running the test added in the [test] commit on the code before the [bugfix] commit.

glozow committed:
MOVEONLY: update_lock_points to txmempool.h

glozow committed:
[bugfix] update lockpoints correctly during reorg

During a reorg, we re-check timelocks on all mempool entries using
CheckSequenceLocks(useExistingLockPoints=false) and remove any
now-invalid entries. CheckSequenceLocks() also mutates the LockPoints
passed in, and we update valid entries' LockPoints using
update_lock_points. Thus, update_lock_points(lp) needs to be called
right after CheckSequenceLocks(lp), otherwise we lose the data in lp.
commit bedf246 introduced a bug by separating those two loops.

glozow commented:
cc @MarcoFalke, thanks for catching my bug :bug:

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -106,8 +106,45 @@ def run_test(self):
 
         self.log.info("Check that the mempool is empty")
         assert_equal(set(self.nodes[0].getrawmempool()), set())
+        assert_equal(set(self.nodes[1].getrawmempool()), set())
```
nit: Why is this needed? The `self.sync_all()` in the next line should take care of that already. I think it should be fine to remove either this line or the next.

glozow replied:
removed

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -106,8 +106,45 @@ def run_test(self):
 
 
+        self.log.info("Simulate natural reorg")
+        self.connect_nodes(0, 1)
```
nit: Why is this needed? Are they not connected already?

glozow replied:
They don't stay in sync if I don't call `connect_nodes` here... it also doesn't work if I call `connect_nodes(1, 0)` instead. So my guess is that this creation of a manual connection from node0 to node1 adds a permission that we need.

MarcoFalke replied:
Ah ok, it is not possible to use compact blocks during IBD

Kejs46 replied:
..

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -106,8 +106,45 @@ def run_test(self):
+        self.log.info("Simulate natural reorg")
+        self.connect_nodes(0, 1)
+        assert_equal(self.nodes[0].getblockcount(), self.nodes[1].getblockcount())
+        assert_equal(set(self.nodes[0].getrawmempool()), set(self.nodes[0].getrawmempool()))
```
nit: Why are the asserts needed? The implicit `self.sync_all` in the next line should assert this already.

glozow replied:
removed

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -106,8 +106,45 @@ def run_test(self):
+        utxo = wallet.get_utxo()
+        timelock_tx_2 = wallet.create_self_transfer(
+            from_node=self.nodes[0],
+            utxo_to_spend=utxo,
```
nit: Why is this needed? The wallet will pick a utxo by itself.

glozow replied:
removed

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -106,8 +106,45 @@ def run_test(self):
+        self.log.info("Disconnect nodes and mine separate forks")
+        self.disconnect_nodes(0, 1)
+        # Don't try to sync nodes because they are disconnected
+        no_sync_func = lambda: True
```
nit: Why is this needed? You can use `self.no_op`

glozow replied:
ahh, didn't know about that - fixed

maflcko reviewed (commented):
Concept ACK. Left some questions on the test.

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -108,6 +108,37 @@ def run_test(self):
+        timelock_tx_2 = wallet.create_self_transfer(
+            from_node=self.nodes[0],
+            mempool_valid=False,
+            locktime=self.nodes[0].getblockcount() + 2
```
nit: missing trailing `,`.

glozow replied:
added

maflcko reviewed (approved):
Too bad this can't properly be tested on regtest. I guess the only way is via a pre-mined main-chain?

Concept ACK bd716bf774d21956ec7ae616d06a2a896568ed18 🙆

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK bd716bf774d21956ec7ae616d06a2a896568ed18 🙆
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjr2gwAqMWws2MpZHhtGe0UDOPIk0PKWrc/Xbg991hVwlMwmHv2ZdgTm0aeYa15
zI0/9PVfKF7P1mTn5iWjJGaoAzsWPyRQAO7imfKwiYOUQnqBBP8s4qpjZxAyE+y4
n9CtBkoDxJd6FBY3/PaC+A/TQCXqrfoefLXMqkCjkFl5aF85YRyAvB69xpKDwMp0
KJLqSga6ThHVkHJOXPoD09yjCEOkT/2K7z2XMTyzGMkzmTBJnhBaLd50Lfy31Ioj
EQz4G7wEskfJkQBATIDzvy0nbwbH38hL30GEVolasAYLUdlxnRQr1QnZEnEUEoV1
MrdzGtQYuBIUyEoaqJyvL/WVb+WEPy1XL/R3VmHf8vWbhAJK2W8RkGW1auE7WZwP
W8HubXwO/dqmXqGd2A9QUaRN1RtYRKL7QwA5ffym++tJQRtpqT4/htuFH4BaIfv/
+Z4d4NPkP8ySa1fhwWvMvLf2JM+5V9572LPInHka1B7t9z9DRjVCN3fb/nMOBzUp
A2NkaSsm
=9LGj
-----END PGP SIGNATURE-----
```


</details>

glozow commented on src/validation.cpp:
```diff
@@ -378,6 +378,8 @@ void CChainState::MaybeUpdateMempoolForReorg(
                 }
```
added

glozow commented on src/txmempool.h:
```diff
@@ -312,6 +312,19 @@ class CompareTxMemPoolEntryByAncestorFee
 };
 
+/**
+ * Update LockPoints in a mempool entry for which the caller only has a const iterator to.
```
removed

naumenkogs commented on src/validation.cpp:
```diff
@@ -378,6 +378,8 @@ void CChainState::MaybeUpdateMempoolForReorg(
             }
         }
+        // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
+        if (!validLP) m_mempool->mapTx.modify(it, update_lock_points(lp));
```
I was thinking "what if we skip those transactions with `should_remove=true` here?". I guess, we'll have the same original problem: they won't be cleared because we would mistakenly think they have valid `lp` in `removeForReorg`?

Perhaps this deserves a comment extension, saying that txs to removed must be updated here as well.

glozow replied:
No, they'll still be cleared - `removeForReorg` just applies `check_final_and_mature` on all entries and removes them if it returns true, it doesn't check whether the `lp` is valid. I'm not 100% sure if we could skip the `should_remove=true` ones, but I think it would be fine. The `CheckSequenceLocks` function doesn't guarantee `lp` to be reliable if it returns false, and the function doesn't even get executed if `CheckFinalTx` returns false.

MarcoFalke replied:
I agree that it is safe to skip if the tx will be removed anyway. When this was introduced in commit 14d6324a248df50cb79fbeb5b60a978687a3b64e it updates unconditionally.

vasild replied:
I was just about to post the same optimization suggestion: change this to `if (!validLP && !should_remove)`.

> I guess, we'll have the same original problem: they won't be cleared because we would mistakenly think they have valid lp in removeForReorg

I don't see it that way - the tx will be removed if this lambda `check_final_and_mature()` returns `true`, so why bother updating that tx data? (I see no reason).

vasild replied:
Sleeping over this, now I think that it is better to leave it as is - just `if (!validLP)` because skipping update for the to-be-removed transactions will spare some CPU cycles but will increase the coupling between `check_final_and_mature()` and `removeForReorg()` - the former will assume the later will remove the transactions and the later must be aware that some transactions have invalid lock points in case it decides to do something else before removal. Plus more complex code. Not worth it.

glozow replied:
> skipping update for the to-be-removed transactions will spare some CPU cycles but will increase the coupling between check_final_and_mature() and removeForReorg()

Agreed 👍

naumenkogs commented:
utACK 28b60ce3122b29ffad2c60790d3bfe9a2ede2471

The bug is pretty scary, good thing we found it this soon, and the fix is straightforward.
The test looks correct.

MarcoFalke commented:
> Too bad this can't properly be tested on regtest. I guess the only way is via a pre-mined main-chain?

Ah no. I don't think this is possible to test at all. The lockpoints are just a cache and if it is outdated, evaluating it will always return false, in which case the lockpoints are re-evaluated. So this is not fixing a behaviour bug, but potentially improving performance a bit?

glozow commented:
> > Too bad this can't properly be tested on regtest. I guess the only way is via a pre-mined main-chain?

> Ah no. I don't think this is possible to test at all. The lockpoints are just a cache and if it is outdated, evaluating it will always return false, in which case the lockpoints are re-evaluated. So this is not fixing a behaviour bug, but potentially improving performance a bit?

The "natural reorg to a new chain that is shorter (but has more work)" is probably not possible without significant changes to regtest chainparams, I think. Since this PR adds assertions for the lockpoints, making sure they're updated is testable. And true, I suppose outdated lockpoints are not _that_ bad, though now I'm wondering if it's worth adding the assertions if it's relatively harmless.

shaavan reviewed (approved):
ACK 28b60ce3122b29ffad2c60790d3bfe9a2ede2471

The updates look good to me, and they seem to fix the bug introduced in #22677.

I was able to run the test successfully on the master branch. Thanks for fixing this quickly, @glozow!

**p.s.** While I ran all the tests on the PR branch using `test/functional/test_runner.py` the following tests failed:

```
feature_coinstatsindex.py                          | ✖ Failed  | 2 s
feature_rbf.py --legacy-wallet                     | ✖ Failed  | 2 s
interface_bitcoin_cli.py                           | ✖ Failed  | 3 s
wallet_transactiontime_rescan.py                   | ✖ Failed  | 6 s

```

Though it is not probably related to this PR. I still wanted to bring it to other contributors' attention. I shall also try running these tests on the master branch to check if the error persists.

Update: The same tests fails with the master too (with top commit d20d6ac545159fb98bd9ac828678ddf20f5d016d). I am not sure what the reason is, though.

MarcoFalke commented:
> Since this PR adds assertions for the lockpoints

I was hoping there was a way to test this on current master without modifying the code, but yeah...

vasild commented on src/txmempool.cpp:
```diff
@@ -649,10 +639,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
-        if (!TestLockPointValidity(chain, lp)) {
-            mapTx.modify(it, update_lock_points(lp));
-        }
+        assert(TestLockPointValidity(chain, it->GetLockPoints()));
```
Is it not better to add this assert as soon as the data is updated? That is, something like this:

<details>
<summary>patch</summary>

```diff
diff --git i/src/txmempool.cpp w/src/txmempool.cpp
index 26cb2ff4b4..8149c6dcd7 100644
--- i/src/txmempool.cpp
+++ w/src/txmempool.cpp
@@ -635,15 +635,12 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
     }
     setEntries setAllRemoves;
     for (txiter it : txToRemove) {
         CalculateDescendants(it, setAllRemoves);
     }
     RemoveStaged(setAllRemoves, false, MemPoolRemovalReason::REORG);
-    for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-        assert(TestLockPointValidity(chain, it->GetLockPoints()));
-    }
 }
 
 void CTxMemPool::removeConflicts(const CTransaction &tx)
 {
     // Remove transactions which depend on inputs of tx, recursively
     AssertLockHeld(cs);
diff --git i/src/validation.cpp w/src/validation.cpp
index 68729e4863..541b7be499 100644
--- i/src/validation.cpp
+++ w/src/validation.cpp
@@ -377,12 +377,15 @@ void CChainState::MaybeUpdateMempoolForReorg(
                     break;
                 }
             }
         }
         // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
         if (!validLP) m_mempool->mapTx.modify(it, update_lock_points(lp));
+        if (!should_remove) {
+            assert(TestLockPointValidity(m_chain, it->GetLockPoints()));
+        }
         return should_remove;
     };
 
     // We also need to remove any now-immature transactions
     m_mempool->removeForReorg(m_chain, check_final_and_mature);
     // Re-limit mempool size, in case we added any transactions
```
</details>

MarcoFalke replied:
Seems better to run it on all remaining txs unconditionally before returning as a sanity check.

glozow replied:
I agree it seems like a better sanity check run at the end of `removeForReorg`, once all of the entries have been processed

vasild commented on test/functional/mempool_reorg.py:
```diff
@@ -108,6 +108,37 @@ def run_test(self):
 
+        self.log.info("Simulate a natural reorg where mempool transaction stays valid but on a different block")
+        self.connect_nodes(0, 1)
+        # Mine a block to create mature coinbase
```
This test passes also if the fix is missing. Is this expected? If yes, then what is the point of the test? I think tests should fail if the bug resurfaces.

On top of this PR:
* `git show b4adc5ad67 |git apply -R`
* recompile
* run `test/functional/mempool_reorg.py` - it passes

MarcoFalke replied:
I don't think it is possible to test this without modifying the source code (e.g. addding an assert) or inspecting the process internal memory.

vasild replied:
So, what is the purpose of this test? It should be removed?

MarcoFalke replied:
I think the goal is to do a reorg without using `invalidateblock`, as above

glozow replied:
The test is supposed to fail without the bugfix if the asserts are there

MarcoFalke replied:
I recall adding the assert and saw this and one other test failing already

vasild replied:
What assert? Something extra, that is not in the PR?

MarcoFalke replied:
Yes, assert.

As mentioned previously it is not possible to test this without modifying the source code or looking at internal memory.

Adding an assert is one way to modify the source code. Another way would be to return the struct over rpc, or log it to the debug log, or ...

glozow replied:
> What assert?

This one: https://github.com/bitcoin/bitcoin/pull/23683/files#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522R642

glozow replied:
In a previous version of this PR I think I had added the assert in a separate commit from the bugfix.

vasild replied:
I see (which assert). So I took latest `master` (7629efcc2c3a8a8a7c17b1300cd466ec6c8c1f3f) and applied this patch:

```diff
--- i/src/txmempool.cpp
+++ w/src/txmempool.cpp
@@ -653,6 +653,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
         if (!TestLockPointValidity(chain, lp)) {
             mapTx.modify(it, update_lock_points(lp));
         }
+        assert(TestLockPointValidity(chain, lp));
     }
 }
```

`test/functional/mempool_reorg.py` then fails because the added assert is hit. This is unmodified `mempool_reorg.py`, without the extension from this PR. If the bug can be reproduced by adding an assert + unmodified `mempool_reorg.py`, then why bother to change (extend) the test?

> I think the goal is to do a reorg in this test without using `invalidateblock`

Why bother if the test fails even without that?

MarcoFalke replied:
reorgs in the real network happen without `invalidateblock`, so I think we should attempt to not use the call

glozow replied:
The new test is not strictly necessary in order to hit this, but it's not pointless. We didn't have any natural reorgs. We used `invalidateblock` as a shortcut to do it manually. I can remove it if you insist, but I don't really see why it's problematic?

vasild replied:
Alright, I agree that for the test the natural reorg is better than `invalidateblock`. But the change in this PR just adds a natural reorg, without removing the `invalidateblock` one. Shouldn't it replace `invalidateblock` with a natural reorg?

I checked the code coverage with and without the extension of the test from this PR (28b60ce3122b29ffad2c60790d3bfe9a2ede2471) and it is the same. In other words, the newly added snippet does not extend the coverage. It exercises code that has already been exercised by preceding parts of `mempool_reorg.py`.

<details>
<summary>commands to compare coverage with/without test extension</summary>

```sh
# compile
F="-fprofile-instr-generate -fcoverage-mapping" CXXFLAGS=${F} LDFLAGS=${F} ./configure and make as usual

# run test and gather coverage data
LLVM_PROFILE_FILE="/tmp/coverage1/%m-%p.profraw" test/functional/mempool_reorg.py
llvm-profdata13 merge /tmp/coverage1/*.profraw -o /tmp/coverage1/all.profdata

# remove topmost commit (test extension)
git show |git apply -R

# run test and gather coverage data
LLVM_PROFILE_FILE="/tmp/coverage2/%m-%p.profraw" test/functional/mempool_reorg.py
llvm-profdata13 merge /tmp/coverage2/*.profraw -o /tmp/coverage2/all.profdata

# compare the coverage (both are identical)
md5 -r /tmp/coverage*/all.profdata
14196d89227bc6028834aeff32b79ff5 /tmp/coverage1/all.profdata
14196d89227bc6028834aeff32b79ff5 /tmp/coverage2/all.profdata
```
</details>

I don't insist on removing the test snippet. I am just trying to understand the point of it. So far, with my understanding, I am ~0 on it because I fail to see what value it adds. If the `invalidate` block reorg is removed and replaced by a natural reorg without reducing code coverage, then that would be a definite +1.

vasild reviewed (commented):
Concept ACK 28b60ce3122b29ffad2c60790d3bfe9a2ede2471

I verified that this PR reverts bedf246f1e2497a3641093c6e8fa11fb34dddac4 modulo some code reorganization that happened in the mean time. Also, it adds an assert that was not present before and a test (that always passes?).

ariard commented:
Concept ACK a64078e

> Ah no. I don't think this is possible to test at all. The lockpoints are just a cache and if it is outdated, evaluating it will always return false, in which case the lockpoints are re-evaluated. So this is not fixing a behaviour bug, but potentially improving performance a bit?

Yes, I don't think this is a bug introducing invalid transactions in the mempool. Just a perf regression.

If you have a reorg from chain H to chain H'' and the lockpoints are not valid anymore on H'', the cache is not going be updated accordingly.

Let's say you have one more reorg from chain H'' to chain H'''. If H''' includes the max `CBlockIndex` (tested in `TestLockPointValidity`), the lockpoints are _anew_ valid. If H''' does not include the max `CBlockIndex`, the lockpoints are going to be recomputed by `CheckSequenceLocks` to determine if the transaction is final.

So if this reasoning is correct, I also agree that caching lockpoints is a perf improvement, in the sense they save few coins fetches and block index walkes ?

vasild commented:
> cc @MarcoFalke, thanks for catching my bug

How did you catch this? It looks pretty obscure to me.

MarcoFalke commented:
> > cc @MarcoFalke, thanks for catching my bug
> 
> How did you catch this? It looks pretty obscure to me.

Through code review, basically. Though, it was a bit tricky because there was never an externally observable logic bug, at most the performance difference. For a few hours I even thought the code could be removed (#23660) :sweat_smile:

jnewbery commented:
Change looks good to me. I agree it fixes the lockpoint invalidity bug introduced in #22677.

A few suggestions for the `check_final_and_mature` lambda:

- A contributing factor to the the bug being introduced was misleading naming (of `update_lock_points()`) and a lack of documentation. Therefore, we should be extra explicit about what's going on in these functions by adding plenty of code documentation.
- the return value is the inverse of what I'd expect. I'd expect a `check_blah()` function to return `true` if the check passes and `false` if it fails.
- the local `should_remove` variable seems unnecessary. The function can just return immediately if one of the checks fails.

I've included a suggested patch below. Feel free to take whatever you think is an improvement.

<details>
<summary>Suggested patch</summary>

```diff
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 26cb2ff4b4..c742548875 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -631,7 +631,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
 
     setEntries txToRemove;
     for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-        if (check_final_and_mature(it)) txToRemove.insert(it);
+        if (!check_final_and_mature(it)) txToRemove.insert(it);
     }
     setEntries setAllRemoves;
     for (txiter it : txToRemove) {
diff --git a/src/validation.cpp b/src/validation.cpp
index f53641e786..8da2e80510 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -350,21 +350,38 @@ void CChainState::MaybeUpdateMempoolForReorg(
     // the disconnectpool that were added back and cleans up the mempool state.
     m_mempool->UpdateTransactionsFromBlock(vHashUpdate);
 
-    const auto check_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
+    // Check whether the transaction is still final/mature, and update
+    // the transaction's cached lockpoints if necessary.
+    // If this returns false, the mempool *must* remove the transaction, since
+    // it is no longer valid for inclusion in the next block (and its cached
+    // lockpoints may be invalid).
+    const auto check_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it) -> bool
         EXCLUSIVE_LOCKS_REQUIRED(m_mempool->cs, ::cs_main) {
-        bool should_remove = false;
         AssertLockHeld(m_mempool->cs);
         AssertLockHeld(::cs_main);
         const CTransaction& tx = it->GetTx();
+
+        // Check transaction finality
+        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) {
+            return false;
+        }
+
+        // Check sequence locks
         LockPoints lp = it->GetLockPoints();
         const bool validLP{TestLockPointValidity(m_chain, lp)};
         CCoinsViewMemPool view_mempool(&CoinsTip(), *m_mempool);
-        if (!CheckFinalTx(m_chain.Tip(), tx, flags)
-            || !CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
-            // Note if CheckSequenceLocks fails the LockPoints may still be invalid
-            // So it's critical that we remove the tx and not depend on the LockPoints.
-            should_remove = true;
-        } else if (it->GetSpendsCoinbase()) {
+
+        if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
+            // Sequence locks fail. The mempool *must* remove this transaction since
+            // we're not going to update its cached lockpoints.
+            return false;
+        } else if (!validLP) {
+            // CheckSequenceLocks has modified lp. Update the cached lockpoints.
+             m_mempool->mapTx.modify(it, update_lock_points(lp));
+        }
+
+        // Check coinbase spend maturity
+        if (it->GetSpendsCoinbase()) {
             for (const CTxIn& txin : tx.vin) {
                 auto it2 = m_mempool->mapTx.find(txin.prevout.hash);
                 if (it2 != m_mempool->mapTx.end())
@@ -373,18 +390,19 @@ void CChainState::MaybeUpdateMempoolForReorg(
                 assert(!coin.IsSpent());
                 const auto mempool_spend_height{m_chain.Tip()->nHeight + 1};
                 if (coin.IsSpent() || (coin.IsCoinBase() && mempool_spend_height - coin.nHeight < COINBASE_MATURITY)) {
-                    should_remove = true;
-                    break;
+                    // Coinbase spend is now immature. Remove tx from mempool.
+                    return false;
                 }
             }
         }
-        // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
-        if (!validLP) m_mempool->mapTx.modify(it, update_lock_points(lp));
-        return should_remove;
+
+        // Transaction is still final and mature. Cached lockpoints have been updated.
+        return true;
     };
 
-    // We also need to remove any now-immature transactions
+    // Remove any transactions that are no longer final/mature, and update cached lockpoints.
     m_mempool->removeForReorg(m_chain, check_final_and_mature);
+
     // Re-limit mempool size, in case we added any transactions
     LimitMempoolSize(
         *m_mempool,
```

</details>

vasild reviewed (approved):
ACK b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b

Notice - the ACKed commit is the `HEAD~` of this PR, without the topmost commit which adds a test which I do not think is harmful and I do not object. But I don't see what value it adds because:

* It passes with and without the fix
* If an [assert](https://github.com/bitcoin/bitcoin/pull/23683#discussion_r765731520) is added to the code to make the test fail, then `mempool_reorg.py` fails even without the extension added in this PR
* The extension from this PR does [not increase the code coverage](https://github.com/bitcoin/bitcoin/pull/23683#discussion_r766553409)

This is not an objection to the added test snippet, rather ~0 on it. I am happy if it gathers enough ACKs from other devs and gets merged.

glozow commented:
> Notice - the ACKed commit is the HEAD~ of this PR, without the topmost commit which adds a test which I do not think is harmful and I do not object. But I don't see what value it adds 

I've removed the test. Perhaps a followup dedicated to test improvements can convert the `invalidateblock`s to natural reorgs.

> A few suggestions for the check_final_and_mature lambda:

Thanks, I've applied (most of) your diff. I see what you mean by the return results being unexpected for something called "check." My wish is to follow the convention of a unary predicate that returns `true` if something needs to be removed. I've renamed check to filter.

vasild commented on src/validation.cpp:
```diff
@@ -350,21 +350,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
+        // So it's critical that we remove the tx and not depend on the LockPoints.
+        if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
+            return true;
+        } else if (!validLP) {
```
nit: `else` is not necessary after `return`. May as well remove it also for consistency with the above `if (!CheckFinalTx...` (which does not include `else` after `return`).

```suggestion
        }
        if (!validLP) {
```

jnewbery replied:
The `!validLP` check here is logically associated with the `!CheckSequenceLocks()` check, so it makes sense to have it in an `else` branch. Logically the two things are the same since the if branch contains an early return, but structuring it as if/else signals to the developer that the two things are connected.

vasild commented on src/validation.cpp:
```diff
@@ -350,21 +350,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
+            return true;
+        } else if (!validLP) {
+            // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
+            m_mempool->mapTx.modify(it, update_lock_points(lp));
```
In a previous incarnation of this PR, this `modify()` would have been executed regardless of the outcome of `CheckFinalTx()` and `CheckSequenceLocks()`. Not so anymore. From https://github.com/bitcoin/bitcoin/pull/23683#discussion_r765732191 I got the impression that it is better to update for all transactions, even the to-be-removed ones.

vasild commented on src/validation.cpp:
```diff
@@ -350,21 +350,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
+            m_mempool->mapTx.modify(it, update_lock_points(lp));
+        }
+        // If the transaction spends any coinbase outputs, check maturity.
+        if (it->GetSpendsCoinbase()) {
```
Previously, this check and the body of the `if` would have been executed before updating the lock points. Now the order is reversed. Could the updated lock points change the outcome of any of this code, e.g. the `mapTx.find()` call?

vasild commented on src/validation.cpp:
```diff
@@ -350,21 +350,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
 
-    const auto check_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
+    // Check whether the transaction is still final and, if it spends a coinbase output, mature.
+    // Update the entry's cached LockPoints if needed. If this returns true, the tx would be invalid
```
> Update the entry's cached LockPoints if needed

This is a bit vague with the current code. The lock points may or may not be updated if this function returns `true` and will be updated for sure if it returns `false`. I think it is better to update them in all cases or change this comment to say "Update the entry's cached LockPoints if false is returned (and also in some cases if true is returned :disappointed:)".

glozow commented:
Sorry for the churn. I had wanted to incorporate the suggestions in https://github.com/bitcoin/bitcoin/pull/23683#issuecomment-989741186 for the lambda, but I now think it's best saved for a different PR.

glozow commented on src/txmempool.cpp:
```diff
@@ -649,10 +639,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
     }
     RemoveStaged(setAllRemoves, false, MemPoolRemovalReason::REORG);
     for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
```
reverted back to original, squashed commit

jnewbery commented:
> Sorry for the churn. I had wanted to incorporate the suggestions in #23683 (comment) for the lambda, but I now think it's best saved for a different PR.

I think it's very important that the comments are updated to clarify what this function is doing. A bug was introduced in this code because the code was unclear, so we should take the opportunity to clarify it to prevent a similar regression being added again.

This branch does fix the recent regression, but I think it's significantly worse than the previous version since good tests and clarifying code comments have been removed.

mzumsande reviewed (commented):
Code Review ACK b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b

I agree that this fixes the "bug", and verified that `mempool_reorg.py` fails if only the assert from b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b is added but not the fix. 
Can't comment on any of the older versions/tests - I think that the refactoring/doc proposed by @jnewbery looks sensible, but could also be a follow-up PR.


Commit message nit:

> During a reorg, we re-check timelocks on all mempool entries using
CheckSequenceLocks(useExistingLockPoints=false) and remove any
now-invalid entries.

That sounds as if we call `CheckSequenceLocks(useExistingLockPoints=false)` for all entries, but we call `TestLockPointValidity()` for all mempool entries, and call 
`CheckSequenceLocks(useExistingLockPoints=false)` only for those with invalid lockpoints.

hebasto commented:
Concept ACK.

hebasto commented on src/txmempool.cpp:
```diff
@@ -649,10 +639,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
     }
     RemoveStaged(setAllRemoves, false, MemPoolRemovalReason::REORG);
     for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-        const LockPoints lp{it->GetLockPoints()};
-        if (!TestLockPointValidity(chain, lp)) {
-            mapTx.modify(it, update_lock_points(lp));
-        }
```
The removed code is no-op actually. After the `if` statement, `it->GetLockPoints()` returns the same value as `lp`.

hebasto commented on src/txmempool.h:
```diff
@@ -312,6 +312,16 @@ class CompareTxMemPoolEntryByAncestorFee
     }
 };
 
+struct update_lock_points
+{
+    explicit update_lock_points(const LockPoints& _lp) : lp(_lp) { }
+
+    void operator() (CTxMemPoolEntry &e) { e.UpdateLockPoints(lp); }
+
+private:
+    const LockPoints& lp;
+};
+
```
Maybe drop `update_lock_points` altogether:
```diff
diff --git a/src/txmempool.h b/src/txmempool.h
index ba1d381a5..f87ecc9cd 100644
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -312,16 +312,6 @@ public:
     }
 };
 
-struct update_lock_points
-{
-    explicit update_lock_points(const LockPoints& _lp) : lp(_lp) { }
-
-    void operator() (CTxMemPoolEntry &e) { e.UpdateLockPoints(lp); }
-
-private:
-    const LockPoints& lp;
-};
-
 // Multi_index tag names
 struct descendant_score {};
 struct entry_time {};
diff --git a/src/validation.cpp b/src/validation.cpp
index 68729e486..a42e1f52a 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -379,7 +379,7 @@ void CChainState::MaybeUpdateMempoolForReorg(
             }
         }
         // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
-        if (!validLP) m_mempool->mapTx.modify(it, update_lock_points(lp));
+        if (!validLP) m_mempool->mapTx.modify(it, [lp](CTxMemPoolEntry& e) { e.UpdateLockPoints(lp); });
         return should_remove;
     };
 
```
?

hebasto reviewed (approved):
ACK b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b

jnewbery commented:
ACK 7b7f53bdbe55389f2b5934dcab1299363d23bcc7

But please do update the comments and add a test in a follow-up.

maflcko reviewed (approved):
re-ACK b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b 🏁

Only changes:
* Remove commit that adds test (bd716bf774)
* Remove useless doxygen comment on struct update_lock_points
* Expand the commit message of commit b4adc5ad67 "[bugfix] update lockpoints correctly during reorg"

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b 🏁

Only changes:
* Remove commit that adds test (bd716bf774)
* Remove useless doxygen comment on struct update_lock_points
* Expand the commit message of commit b4adc5ad67 "[bugfix] update lockpoints correctly during reorg"
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjyJgwAiIcxIULHGXbeJdpJrr/yTOIL9za7KmSpiwuFZgB5alGaxZHVwNXz5tMs
JgREgzSCQyGFl6otzezJKQ4ZwZ3+m6Xkjlop9DGQNgVXlRPs0hAUwePLj0c1a0Y7
XpJaCS8U3Ovi/av/eTXBaAAVZ/tH/ngqiSo1Y7E41e4c12IGtE4JlgXaaI8+vaC8
4oF29XugN/FWuoY8WXqXGHwnRyxyhid9QR7uLfmQC+0tcEUW2bu9yDH+hWuFeNs1
8gu6mb76OuMqRQR/PRA7Jx+vN2WlN9eCua+TFVWvxS9vl06jOOrz2Yw4qC8gisgz
J5rzAH8y5aMnztO9I2//smKojsxtierk2dIe2BXBGsgoo8GqxsbLocCHuTrvlt56
eBJuUZLFYc3G5WtxD55cc6WpPGMVJJ/5sfptnYYNiaaca6bIYk1dmBklAQDdeyGg
GtwDfvZIbYFMU2Aq+cAta59oo7Xtw9tSWS6Kp1sKJOQxQfhB39XBxjgc5qwZ7L1z
hSX5amzs
=W5Xo
-----END PGP SIGNATURE-----
```


</details>
